### DISCUSSION TITLE
Form with browser steals focus from window on top

### DISCUSSION THREAD

**Discussion by assopri** [cefsharp/CefSharp#3599] [Q&A]
In my app I create winforms cefsharp browsers outside of the main thread. I use the following code to create form for hosting them (each separate browser is hosted in separate tabpage of tabcontrol in the form):

`private void CreateForm()
        {
            Form = new Form1( manualResetEvent);
            Form.ProceedButtonClicked += (_, browser) =>
            {
                browserWrapperDictionary[browser].WaitingForUserAction = false;
            };

            _thread = new Thread(() =>
            {
                Application.EnableVisualStyles();
                Application.SetCompatibleTextRenderingDefault(true);

                while (!_cancellationTokenSource.IsCancellationRequested)
                {
                    if (Form.ShowFormFlag) { Form.ShowFormFlag = false; Form.Show(); }
                    if (Form.HideFormFlag) { Form.HideFormFlag = false; Form.Hide(); }
                    Application.DoEvents();
                   
                    Thread.Sleep(100);
                }

                Form.Close();
                Form.Dispose();
                
            });

            _thread.SetApartmentState(ApartmentState.STA);
            _thread.Start();
        }`

However I found that when another browser is created form with browser steals focus from my main form and appears to be on top of this form. How can I avoid that?

My OS is Windows7 64x , VS 2019 installed. I use Winforms Cefsharp version, .NET 4.5.2 . Cef version is 90.6.50.0

**Comment by amaitland** [comment]
> However I found that when another browser is created form with browser 

Can you clarify what you mean by this?

**Reply by assopri** [reply]
I guess I'm having the same issue, as described here : https://github.com/cefsharp/CefSharp/issues/2928 . I tried solution, which was offered there https://github.com/cefsharp/CefSharp/issues/2928#issuecomment-546772753 , however on my 90.6.50 version I get the following error:
![image](https://user-images.githubusercontent.com/73235464/126487049-26afa248-092a-4007-b515-cbc392e457f5.png)

To be short - InvokeOnUiThreadIfRequired is not available. Any workaround?

**Reply by amaitland** [reply]
The old focus handler behaviour was restored as per https://github.com/cefsharp/CefSharp/issues/2928#issuecomment-609515408

If you are using multiple forms then make sure you are using an `ApplicationContext` as detailed in https://github.com/cefsharp/CefSharp/issues/2928#issuecomment-626303046

> To be short - InvokeOnUiThreadIfRequired is not available. Any workaround?

It's simply a helper method that wraps [Control.BeginInvoke](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.control.begininvoke?view=net-5.0). See https://github.com/cefsharp/CefSharp/blob/cefsharp/75/CefSharp.WinForms/Internals/ControlExtensions.cs#L20 for example.

**Reply by assopri** [reply]
Thank you for reply! I have a little bit different code organization. I have one form containing several browsers (one per tab). Form itself is created and managed with the following code:

`private void CreateForm()
        {
            Form = new Form1( manualResetEvent);
            Form.ProceedButtonClicked += (_, browser) =>
            {
                browserWrapperDictionary[browser].WaitingForUserAction = false;
            };

            _thread = new Thread(() =>
            {
                Application.EnableVisualStyles();
                Application.SetCompatibleTextRenderingDefault(true);

                while (!_cancellationTokenSource.IsCancellationRequested)
                {
                    if (Form.ShowFormFlag) { Form.ShowFormFlag = false; Form.Show(); }
                    if (Form.HideFormFlag) { Form.HideFormFlag = false; Form.Hide(); }
                    Application.DoEvents();
                   
                    Thread.Sleep(100);
                }

                Form.Close();
                Form.Dispose();
                
            });

            _thread.SetApartmentState(ApartmentState.STA);
            _thread.Start();
        }`

![image](https://user-images.githubusercontent.com/73235464/126604589-7869944b-f982-412d-8c2e-30b47eb1334d.png)


Loading webpage in each particular tab (which is supposed to repeat many times) steals focus from the main application (which creates the form, BUT not from another applications launched). What I cannot understand right now is how I can apply ApplicationContext approach to my case. Thank you for your assistance!

**Reply by amaitland** [reply]
What version of CefSharp are you using?

**Reply by assopri** [reply]
> What version of CefSharp are you using?

90.6.50

**Reply by amaitland** [reply]
Which form is the `ChromiumWebBrowser` instance in? 

Is there a reason you are creating a new `Thread` to run your form? Rarely is this a good idea.

**Reply by assopri** [reply]
If I create form right from my thread (which is managing load and other browser interaction) form does not process it's event loop (because there is no separate thread to process it), which (as far as I understand) is necessary to execute url loading (in webrowser located in form) and other browser interaction. Note, that form with browser is not main form of my appliation (and moreover - I cannot use my main UI thread to process its event loop).

![image](https://user-images.githubusercontent.com/73235464/127115381-921e2cff-a450-42e0-bc87-cc71611bc9e0.png)

**Reply by amaitland** [reply]
CefSharp doesn't require a separate thread for UI. It should ideally be run in your applications main thread. Using a second UI thread is in almost all cases unnessicary. 

You would typically call Application.Run to run the WinForms message loop.

https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.application.run?view=net-5.0

For examples of using an ApplicationContext see

- https://github.com/cefsharp/CefSharp.MinimalExample/commit/98830e03e80073acd4e509b4ce2a1fcece21a6d7#diff-a910132c7c8b09d7e87360acdd52f409R8
- https://stackoverflow.com/a/13406508/4583726

A screenshot of some code unfortunately doesn't tell me much.

If you fork the minimum example, the minimum changes required to reproduce the problem and push your changes to GitHub I'll have a look. 

https://github.com/cefsharp/CefSharp.MinimalExample

**Reply by assopri** [reply]
Kindly appreciate your help! Here is my basic example:
https://github.com/assopri/CefSample

Here I create new form in thread https://github.com/assopri/CefSample/blob/210fbf1a5eac73cc0b420a4880ea2f8e1f990741/CefBrowserWrapper/CefBrowserWrapperBase.cs#L153 .

In my real application it can happen that I have created 5-10 forms with 5-10 browsers in each. So I assume that if I don't use a separate thread for processing their UI, I will get delays in interaction with my main form.

Is ApplicationContext approach appropriate in my case? How I can use however leave browser form UI processing for the secondary threads?

**Reply by amaitland** [reply]
> Kindly appreciate your help! Here is my basic example:
https://github.com/assopri/CefSample

Is there a reason you use Task.Wait() instead of using async/await? 

> So I assume that if I don't use a separate thread for processing their UI, I will get delays in interaction with my main form.

By default CEF already runs it's own message loop in a separate thread. From what I can tell your code constantly calls Task.Wait() blocking the UI thread. You just need to switch from blocking every async call to using async/await or ContinueWith and you should be able to use a single UI thread. 

> Is ApplicationContext approach appropriate in my case? How I can use however leave browser form UI processing for the secondary threads?

To use an ApplicationContext you would need to use a single UI thread.

**Reply by assopri** [reply]
If I will invest a lot of time in changing my app call hierarchy to use async instead of Wait - how in this case I can use ApplicationContext ? How exactly should I change form creation to use it? https://github.com/assopri/CefSample/blob/210fbf1a5eac73cc0b420a4880ea2f8e1f990741/CefBrowserWrapper/CefBrowserWrapperBase.cs#L153

**Reply by amaitland** [reply]
> how in this case I can use ApplicationContext ? How exactly should I change form creation to use it?

I had a look at your example in detail and you are creating the forms separately, so the `ApplicationContext` may not be required.

I've reduced your example to the absolute minimum to reproduce the issue https://github.com/amaitland/CefSample/tree/demo/textbox

From what I can tell the problem with form stealing focus is actually related to the `TabControl`. Switching to a `TextBox` instead of ` ChromiumWebBrowser` and the problem still occurs.

https://github.com/amaitland/CefSample/blob/demo/textbox/CefBrowserWrapper/BrowserForm.cs#L19

Others are reporting similar issues 
- https://stackoverflow.com/questions/3478058/net-winforms-tabcontrol-steals-focus-on-form-activation
- https://www.codeproject.com/Questions/1243997/Winform-tabcontrol-unexpected-focus-change-upon-sw

You can look at the `TabControl` source via https://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/TabControl.cs
Might give you some clues on how to workaround the issue. Otherwise you just manually refocus your form.

**Reply by assopri** [reply]
Its awesome! You are completely right! That is about tabcontrol. When I remove changing selected tab "tabControl1.SelectedTab = (browser.Parent as TabPage);" and hiding tabs - focus remains where it must be. However - when avoid selecting tab with active browser I have another issue - webpages (in all browsers except for one in selected tab) - are not loading, so I get timeout exception here:

![image](https://user-images.githubusercontent.com/73235464/127441268-6954c997-a969-463f-9700-7f151e95e3be.png)

any workarounds?

p.s. how can I thank you for localization problem with the tabcontrol?

**Reply by amaitland** [reply]
Sorry for the delayed response, I've been sick most of the last week.

The underlying browser will only be created when the Handle is created. You should be able to set the size of the ChromiumWebBrowser instance and call ChromiumWebBrowser.CreateControl to explicitly load the control.

https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.control.createcontrol?view=net-5.0

> p.s. how can I thank you for localization problem with the tabcontrol?

Your support via GitHub sponsors is sufficient thanks 👍

**Reply by assopri** [reply]
Hello! I hope you are well!
